examples/reshape.cpp

```diff
@@ -0,0 +1,728 @@
+ *  i.e. A = [A^{0}_{00}, A^{0}_{01}, A^{0}_{02},
+ *            A^{0}_{10}, A^{0}_{11}, A^{0}_{12},
+ *            A^{0}_{20}, A^{0}_{21}, A^{0}_{22},
+ *            A^{1}_{00}, A^{1}_{01}, A^{1}_{02},
+ *            A^{1}_{10}, A^{1}_{11}, A^{1}_{12},
+ *            A^{1}_{20}, A^{1}_{21}, A^{1}_{22}];
+ *
+ *  Layout 2:
+ *  Matrix entries are first ordered by matrix number,
+ *  then by column number, and finally by row number.
+ *  i.e. A = [A^{0}_{00}, A^{1}_{00}, A^{0}_{01},
+ *            A^{1}_{01}, A^{0}_{02}, A^{1}_{02},
+ *            A^{0}_{10}, A^{1}_{10}, A^{0}_{11},
+ *            A^{1}_{11}, A^{0}_{12}, A^{1}_{12},
+ *            A^{0}_{20}, A^{1}_{20}, A^{0}_{21},
+ *            A^{1}_{21}, A^{0}_{22}, A^{1}_{22}];
+ *
+ * The extension to N > 2 matrices follows by direct
+ * extension. By exploring different data layouts,
+ * we can assess which performs best under a given
+ * execution policy and architecture.
+ *
+ *  RAJA features shown:
+ *    - RAJA::forall kernel execution method
+ *    - RAJA::View
+ *    - RAJA::Layout
+ *    - RAJA::make_permuted_layout method
+ *
+ * If CUDA is enabled, CUDA unified memory is used.
+ */
+
+/*
+ * Define number of threads in a GPU thread block
+ */
+#if defined(RAJA_ENABLE_CUDA)
+constexpr int CUDA_BLOCK_SIZE = 256;
+#endif
+
+#if defined(RAJA_ENABLE_HIP)
+constexpr int HIP_BLOCK_SIZE = 256;
+#endif
+
+//
+//Function for checking results
+//
+template <typename T>
+void checkResult(T C, int nMat, int nRows, int nCols);
+
+int main(int RAJA_UNUSED_ARG(argc), char **RAJA_UNUSED_ARG(argv[]))
+{
+
+  std::cout << "\n\nRAJA batched matrix multiplication exercise...\n";
+
+// Dimensions of matrices
+  constexpr int N_c = 3;
+  constexpr int N_r = 3;
+
+// Number of matrices
+  constexpr int N = 8000000;
+
+// Number of iterations
+  constexpr int NITER = 20;
+
+  std::cout << "\n Number of matrices to be multiplied: " << N << " \n \n";
+
+//
+// Initialize a RAJA timer object
+// and variable to store minimum run time
+//
+  auto timer = RAJA::Timer();
+  double minRun = std::numeric_limits<double>::max();
+
+//
+// Allocate space for data in layout 1
+//
+  double *A = memoryManager::allocate<double>(N_c * N_r * N);
+  double *B = memoryManager::allocate<double>(N_c * N_r * N);
+  double *C = memoryManager::allocate<double>(N_c * N_r * N);
+
+//
+// Layout 1
+//
+// make_permuted_layout takes the number of entries in each dimension and a
+// templated array indicating index arguments with slowest to fastest stride.
+// Standard C++ arrays are used to hold the number of entries in each component.
+// This example uses double braces to initalize the array and its subobjects.
+// The layout object will index into the array as the following C macro would
+// #define Aview(e, r, c) A[c + N_c*(r + N_r*e)].
+//
+// RAJA::Layout objects may be templated on dimension, argument type, and
+// index with unit stride. Here, the column index has unit stride (argument 2).
```

Comment:
Can you supply the permutation at compile time and have it produce a view with the right unit stride index?

Reply:
yes, I pushed a version for that just now. Let me know what you think.